src/main/scala/com/actian/spark_vector/colbuffer/ColumnBuffer.scala

```diff
@@ -22,89 +22,173 @@ import com.actian.spark_vector.colbuffer.singles._
+  val valueTypeOut = classTag[OUT]
+
+  @inline def put(source: IN, buffer: ByteBuffer): Unit
+
+  @inline def get(buffer: ByteBuffer): OUT
+}
+
+private[colbuffer] sealed abstract class RWColumnBuffer(col: ColumnBuffer[_, _])
+  extends Serializable {
+  protected final val NullMarker = 1: Byte
+  protected final val NonNullMarker = 0: Byte
+
+  val alignSize = col.alignSize
+  val nullable = col.nullable
+  val maxValueCount = col.maxValueCount
+
+  def position(): Int
+
+  def clear(): Unit
+}
 
-  val valueType = classTag[T]
-  val values = ByteBuffer.allocateDirect(maxValueCount * valueWidth).order(ByteOrder.nativeOrder())
-  val markers = ByteBuffer.allocateDirect(maxValueCount).order(ByteOrder.nativeOrder())
-  private val nullValue = Array.fill[Byte](alignSize)(0: Byte)
+class WriteColumnBuffer[@specialized T: ClassTag](col: ColumnBuffer[T, _]) extends RWColumnBuffer(col) {
+  val values = ByteBuffer.allocateDirect(col.maxValueCount * col.valueWidth).order(ByteOrder.nativeOrder())
+  private val nullValue = Array.fill[Byte](col.alignSize)(0: Byte)
+  lazy val markers = ByteBuffer.allocateDirect(col.maxValueCount).order(ByteOrder.nativeOrder())
 
-  protected def put(source: T, buffer: ByteBuffer): Unit
+  /** Make valueType visible outside */
+  val valueType = col.valueTypeIn
 
+  @throws(classOf[BufferOverflowException])
   def put(source: T): Unit = {
-    put(source, values)
-    if (nullable) {
+    col.put(source, values)
+    if (col.nullable) {
       markers.put(NonNullMarker)
     }
   }
 
-  @throws(classOf[IllegalArgumentException])
+  @throws(classOf[IllegalStateException])
+  @throws(classOf[BufferOverflowException])
   def putNull(): Unit = {
-    if (!nullable) {
-      throw new IllegalArgumentException(s"Cannot store NULL values in non-nullable '${name}' column.")
-    }
+    if (!col.nullable) throw new IllegalStateException(s"Cannot store NULLs in non-nullable column '${col.name}'.")
     markers.put(NullMarker)
     values.put(nullValue)
   }
 
-  def size: Int = {
+  override def position(): Int = {
     var ret = values.position()
-    if (nullable) {
+    if (col.nullable) {
       ret += markers.position()
     }
     ret
   }
 
   def flip(): Unit = {
     values.flip()
-    if (nullable) {
+    if (col.nullable) {
       markers.flip()
     }
   }
 
-  def clear(): Unit = {
+  override def clear(): Unit = {
     values.clear()
-    if (nullable) {
+    if (col.nullable) {
       markers.clear()
     }
   }
 }
 
-/** Trait to be used when implementing a companion object for a typed ColumnBuffer
+class ReadColumnBuffer[@specialized T: ClassTag](col: ColumnBuffer[_, T]) extends RWColumnBuffer(col) {
+  private val values = Array.ofDim[T](col.maxValueCount)
+  private val isNullValue = Array.ofDim[Boolean](col.maxValueCount)
+  private lazy val markers = Array.ofDim[Byte](col.maxValueCount)
+
+  private var left = 0
+  private var right = 0
+
+  /** Make valueType visible outside */
```

Comment:
this will go in the scala docs. Better describe what it is, i.e. the classTag of T.